src/scaler/mod.rs

```diff
@@ -39,18 +60,363 @@ pub trait Scaler {
+    cleanup_timeout: std::time::Duration,
+}
+
+impl<T, P> BackoffAwareScaler<T, P>
+where
+    T: Scaler + Send + Sync,
+    P: Publisher + Send + Sync + 'static,
+{
+    pub fn new(scaler: T, notifier: P, notify_subject: &str, model_name: &str) -> Self {
+        Self {
+            scaler,
+            notifier,
+            notify_subject: notify_subject.to_owned(),
+            model_name: model_name.to_string(),
+            expected_events: Arc::new(RwLock::new(Vec::new())),
+            event_cleaner: Mutex::new(None),
+            // This is hardcoded for now but could be adjusted based on the scaler in the future if needed
+            cleanup_timeout: std::time::Duration::from_secs(30),
+        }
+    }
+
+    pub async fn event_count(&self) -> usize {
+        self.expected_events.read().await.len()
+    }
 
-    /// An optional to implement function that can tell the scaler to enter a backoff period. When
-    /// in backoff, a scaler should always return an empty `Vec<Command>` for all operations. Please
-    /// be careful when calling backoff until you have sent the commands to their intended location
-    /// as otherwise the might not rerun until they are reconciled again.
+    /// Adds events to the expected events list
     ///
-    /// This function should return as quickly as possible, spawning a task that can switch the
-    /// state back from backoff. The passed in channel is meant to be used to notify the caller that
-    /// the backoff is complete with the name of the model. This name can then be used by the caller
-    /// to run a reconciliation pass to ensure everything is done. This is intended to workaround
-    /// the fact that there is no way to tell when we finished stopping/starting actors
+    /// # Arguments
+    /// `events` - A list of (success, failure) events to add to the expected events list
+    /// `clear_previous` - If true, clears the previous expected events list before adding the new events
+    async fn add_events<I>(&self, events: I, clear_previous: bool)
+    where
+        I: IntoIterator<Item = (Event, Option<Event>)>,
+    {
+        let mut expected_events = self.expected_events.write().await;
+        if clear_previous {
+            expected_events.clear();
+        }
+        expected_events.extend(events);
+        self.set_timed_cleanup().await;
+    }
+
+    /// Removes an event pair from the expected events list if one matches the given event
+    /// Returns true if the event was removed, false otherwise
+    async fn remove_event(&self, event: &Event) -> Result<bool> {
+        let mut expected_events = self.expected_events.write().await;
+        let before_count = expected_events.len();
+        expected_events.retain(|(success, fail)| {
+            !evt_matches_expected(success, event)
+                && !fail
+                    .as_ref()
+                    .map(|f| evt_matches_expected(f, event))
+                    .unwrap_or(false)
+        });
```

Comment:
This is hard to follow. Are we removing if the event matches _either_ the success or fail in any of the `expected_events`?

Reply:
Yes. Any command we send out could have either a success case or an optional error case. Matching either one means we found one of the expected events

Reply:
It took me a minute to grok what was happening. Mind adding a comment there?